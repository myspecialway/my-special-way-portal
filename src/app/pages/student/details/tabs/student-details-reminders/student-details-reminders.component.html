
<div *ngIf="student" class="msw-student-details-reminders">
  <div class="msw-student-header-nav-name" *ngIf="!isNewStudent && student">
    {{ student.firstname }} {{ student.lastname }}
  </div>
  <mat-card *ngFor="let reminder of student.reminders" class="student-reminders selected-radius"
            (click)="showReminderDialog(reminder)">
    <div class="table">
      <header class="row">
        <div class="cell reminder-type">{{ reminderType[reminder.type] }}</div>
        <div class="cell mat-slide-toggle">
          <mat-slide-toggle
            [(ngModel)]="reminder.enabled"
            (click)="$event.stopPropagation()"
            (ngModelChange)="toggleActivity(reminder)"></mat-slide-toggle>
        </div>
      </header>
      <div class="empty-dialog-message" *ngIf="!reminder.schedule.length">
        <p>{{ emptyDialogString }}</p>
      </div>
      <div *ngFor="let slot of reminder.schedule" class="row">
        <div class="cell days">{{ getSelectedDays(slot) }}</div>
        <div class="cell hours">{{ getSelectedHours(slot) }}</div>
      </div>
    </div>
  </mat-card>
</div>

