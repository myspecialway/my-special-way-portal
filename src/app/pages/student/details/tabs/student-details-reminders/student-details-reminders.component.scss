.msw-student-details-reminders {
  display: flex;
  align-items: center;
  flex-direction: column;
  padding: 48px;
  & > .student-reminders {
    margin: 0 0 30px 0;
  }
  .student-reminders {
    width: 600px;
    min-height: 120px;
    transition: opacity 0.15s ease-in-out;
  }
  .mat-slide-toggle {
    float: left;
  }
  ::ng-deep .mat-slide-toggle.mat-checked:not(.mat-disabled) .mat-slide-toggle-bar {
    background-color: #f4e4ff !important;
  }
  ::ng-deep .mat-slide-toggle.mat-checked:not(.mat-disabled) .mat-slide-toggle-thumb {
    background-color: #8b66a7 !important;
  }
  .mat-card {
    padding: 0;
  }
  div.row:first-child p {
    font-size: 24px;
  }
  div.row:nth-child(2) .cell p,
  div.row:nth-child(3) .cell p {
    font-size: 16px;
    font-weight: 500;
  }
  .row {
    line-height: 0.7;
    display: table-row;
    .value {
      text-align: left;
      font-size: 14px !important;
    }
  }
  .table {
    width: 100%;
    display: table;
    text-align: right;
    @media screen and (max-width: 580px) {
      display: block;
    }
  }
  .cell {
    padding: 6px 12px;
    display: table-cell;
    @media screen and (max-width: 580px) {
      padding: 2px 16px;
      display: block;
    }
  }
  .selected-radius:hover {
    border: 1px solid #8b66a7;
    border-radius: 5px;
    opacity: 1;
  }
}
