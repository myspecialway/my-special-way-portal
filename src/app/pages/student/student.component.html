<div data-test-id="students-page" class="main-content msw-student-container">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">תלמידים</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-new-student-button" mat-raised-button color="primary" routerLink="/student/_new_"  *ngIf="mayAddStudent">
          <i class="material-icons">
            add
          </i>
          הוסף תלמיד חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table [dataSource]="dataSource" class="msw-table" matSort matSortDisableClear>
          <ng-container matColumnDef="studentName">
            <mat-header-cell *matHeaderCellDef>שם</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{ row.firstname }} {{ row.lastname }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="gradeId">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{(row.class && row.class.name) || '' }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" class="username" [routerLink]="['/student', row._id]">{{ row.username }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [attr.data-test-id]="'delete-user-button-' + row.username" [disabled]="!mayDeleteStudent"
                      (click)="deleteStudent(row._id, row.firstname, row.lastname, row.gradeId, row.gender)"><i class="material-icons"> delete </i> </button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
