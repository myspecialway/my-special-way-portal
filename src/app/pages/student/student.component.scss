@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.msw-student-container .mat-cell:last-child,
.msw-student-container .mat-header-cell:last-child {
  flex-grow: 0;
  flex-basis: 100px;
}
.msw-student-container .mat-header-cell {
  font-weight: bold;
}

.mat-column-deleteUser button {
  display: none;
}
.mat-row:hover {
  .mat-cell {
    color: #8c37cd;
  }
  .mat-column-deleteUser button {
    display: block;
  }
}
