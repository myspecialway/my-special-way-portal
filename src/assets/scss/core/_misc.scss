body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
