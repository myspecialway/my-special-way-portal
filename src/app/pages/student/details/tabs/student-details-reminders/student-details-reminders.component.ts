import { REMINDERS_CONSTANTS, IDbReminderTime } from './../../../../../models/reminder-time.model';
import { StudentService } from './../../../services/student.service';
import { ReminderType, IReminder } from './../../../../../models/reminder.model';
import { first } from 'rxjs/operators';
import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { SubscriptionCleaner } from '../../../../../decorators/SubscriptionCleaner.decorator';
import { MatDialog } from '@angular/material';
import { AddStudentReminderDialogComponent } from '../../../dialogs/reminders/add/add-student-reminder.dialog';
import Student from '../../../../../models/student.model';
import to from 'await-to-js';
import * as _ from 'lodash';

@Component({
  selector: 'app-student-details-reminders',
  templateUrl: './student-details-reminders.component.html',
  styleUrls: ['./student-details-reminders.component.scss'],
})
export class StudentDetailsRemindersComponent implements OnInit {
  idOrNew: string;

  student: Student;
  dayNames: string[] = Array.from(REMINDERS_CONSTANTS.days);
  protected reminderType = ReminderType;
  emptyDialogString = 'בחר יום ושעה';

  @SubscriptionCleaner()
  subCollector;

  constructor(private route: ActivatedRoute, public dialog: MatDialog, private studentService: StudentService) {}

  ngOnInit() {
    if (this.route && this.route.parent) {
      this.subCollector.add(
        this.route.parent.params.subscribe(async (params) => {
          // this.idOrNew = params.idOrNew;
          await this.fetchStudent(params.idOrNew);
        }),
      );
    }
  }

  showReminderDialog(reminder: IReminder) {
    const dialogRef = this.dialog.open(AddStudentReminderDialogComponent, {
      data: reminder,
      height: '376px',
      width: '631px',
    });

    this.subCollector.add(
      dialogRef
        .afterClosed()
        .pipe(first())
        .subscribe(async (data: IReminder) => {
          await this.updateStudentReminders(data);
        }),
    );
  }

  async updateStudentReminders(data: IReminder) {
    if (!data) return;
    const { class: _class, reminders, ...studentQuery } = this.student;
    const class_id = _class ? _class._id : '';
    const newReminders = reminders.map((_reminder) => {
      if (_reminder.type === data.type) {
        return data;
      }
      return _reminder;
    });

    const [err] = await to<string>(this.studentService.update({ ...studentQuery, reminders: newReminders, class_id }));
    if (!err) {
      await this.fetchStudent(studentQuery._id.toString());
    }
  }

  async fetchStudent(id: string) {
    const [err, student] = await to<Student>(this.studentService.getById(id));
    if (!err) {
      this.student = _.cloneDeep(student as Student);
    }
  }

  getSelectedDays(slot: IDbReminderTime): string {
    if (!slot.daysindex) {
      return '';
    }
    slot.daysindex.sort();
    return slot.daysindex.map((dayIndex) => this.dayNames[dayIndex]).join(',');
  }

  getSelectedHours(slot: IDbReminderTime): string {
    if (!slot.hours) {
      return '';
    }
    slot.hours.sort();
    return slot.hours.join(',');
  }

  toggleActivity(reminder: IReminder) {
    this.updateStudentReminders(reminder);
  }
}
