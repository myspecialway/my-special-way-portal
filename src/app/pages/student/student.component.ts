import { first } from 'rxjs/operators';
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { DeleteStudentDialogComponent } from './dialogs/delete/delete-student.dialog';
import { StudentService } from './services/student.service';
import Student from '../../models/student.model';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';
@Component({
  selector: 'app-student',
  templateUrl: './student.component.html',
  styleUrls: ['./student.component.scss'],
})
export class StudentComponent implements OnInit {
  displayedColumns = ['studentName', 'gradeId', 'username', 'deleteUser'];
  dataSource = new MatTableDataSource<Student>();

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  @SubscriptionCleaner()
  subCollector;

  constructor(private studentService: StudentService, private dialog: MatDialog) {}

  ngOnInit() {
    this.dataSource.sort = this.sort;

    this.subCollector.add(
      this.studentService.getAllStudents().subscribe((data) => {
        this.dataSource.data = [...data];
      }),
    );
  }

  async deleteStudent(id: number, firstName: string, lastName: string, gradeId: string) {
    const dialogRef = this.dialog.open(DeleteStudentDialogComponent, {
      data: { id, firstName, lastName, gradeId },
    });

    dialogRef
      .afterClosed()
      .pipe(first())
      .subscribe(async (deletionConfirmed) => {
        if (!deletionConfirmed) {
          return;
        }

        try {
          await this.studentService.delete(id);
        } catch (error) {
          // TODO: implement error handling on UI
          console.error('Error handling not implemented');
          throw error;
        }
      });
  }
}
