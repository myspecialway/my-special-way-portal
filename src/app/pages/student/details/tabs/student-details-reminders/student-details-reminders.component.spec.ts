import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Platform } from '@angular/cdk/platform';
import { Observable } from 'rxjs-compat';
import { ActivatedRoute, Router, RouterModule } from '@angular/router';
import { StudentDetailsRemindersComponent } from './student-details-reminders.component';
import { StudentDetailsComponent } from '../../student-details.component';
import { MatDialog } from '@angular/material';
import { StudentService } from '../../../services/student.service';
let studentReminderDialogMock: Partial<MatDialog>;

describe('Student Details Reminders Component', () => {
  let studentServiceMock: Partial<StudentService>;

  beforeEach(async () => {
    studentReminderDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: jest.fn().mockReturnValue(Observable.of(true)),
      }),
    };

    studentServiceMock = {
      // getById: jest.fn().mockReturnValue(Promise.resolve({})),
      update: jest.fn(),
      getById: jest.fn(),
    };

    TestBed.configureTestingModule({
      imports: [RouterModule.forRoot([])],
      declarations: [StudentDetailsComponent, StudentDetailsRemindersComponent],
      providers: [
        {
          provide: Router,
          useValue: {
            forRoot: jest.fn(),
          },
        },
        Platform,
        {
          provide: ActivatedRoute,
          useValue: {
            parent: {
              params: Observable.of({ idOrNew: '_new_' }),
            },
          },
        },
        { provide: StudentService, useValue: studentServiceMock },
        { provide: MatDialog, useValue: studentReminderDialogMock },
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  describe('with _new_ student path', () => {
    beforeEach(async () => {});

    it('should render the component as described in snapshot', () => {
      const fixture = TestBed.createComponent(StudentDetailsRemindersComponent);
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });

    it('should subscribe to params onInit', () => {
      const activatedRouteMock = {
        parent: {
          params: {
            subscribe: jest.fn().mockImplementationOnce((callback) => callback({})),
          },
        },
      } as never;
      const studentReminderDialogInitMock = {
        parent: {
          params: {
            subscribe: jest.fn().mockImplementationOnce((callback) => callback({})),
          },
        },
      } as never;
      const studentDetailsReminder = new StudentDetailsRemindersComponent(
        activatedRouteMock,
        studentReminderDialogInitMock,
      );

      // when
      studentDetailsReminder.ngOnInit();

      // then
      // tslint:disable-next-line:no-non-null-assertion test case only
      const subscribeMock = (activatedRouteMock as ActivatedRoute).parent!.params.subscribe as jest.Mock;
      expect(subscribeMock).toHaveBeenCalled();
    });
  });
});
