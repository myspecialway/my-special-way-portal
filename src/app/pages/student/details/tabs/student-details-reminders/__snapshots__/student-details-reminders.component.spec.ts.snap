// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Details Reminders Component with _new_ student path should render the component as described in snapshot 1`] = `
<app-student-details-reminders
  dayNames={[Function Array]}
  dialog={[Function Object]}
  medicineToggleMode="false"
  rehabToggleMode="false"
  reminderType={[Function Object]}
  route={[Function Object]}
  studentService={[Function Object]}
  subCollector={[Function Subscription]}
>
  
</app-student-details-reminders>
`;
